articles/getting-started-with-spring-websockets/index.md

```diff
@@ -0,0 +1,414 @@
+    let message = input.value;
+    
+    client.send('/app/chat', {}, JSON.stringify({message: message}));
+
+}
+```
+
+That might be quite a bit to unpack so let’s go step by step. First, we create a sockjs client instance with the URL where the handshake will take place. Then we create a stomp client instance using our sockjs client which we will use to connect to the server. With this stomp client instance, we call the `connect` method with an empty object and a callback function. The empty object represents the headers we will send with our frame. In the callback function, we subscribe to receive messages at our `/topic/messages` endpoint. When we subscribe we provide another callback that will take the frame from the server and put the message contents on the HTML page. Next, we have our `sendMessage` function. This simply takes the value of our input element and sends it to the `/app/chat` endpoint with empty headers.
+
+Now if you test out our client, you will notice every time you press send it will add the text to the screen. More importantly, try opening it up on two windows side by side. You will see that both windows will receive the message.
+
+### Securing Our Endpoints
+
+One important flaw with our back-end is that anyone can connect to our server whether we know them or not. We need to add authentication to our endpoints so that only valid users can send messages. As far as I know, there isn’t a particular standard for securing WebSockets since it is still a relatively new technology. Yet, I have found a reasonably effective way to secure our endpoints that you could use. Before we start coding there are a couple of things to take note of. The javascript WebSocket libraries don’t allow you to add Authorization headers to the handshake. So instead, we will authenticate the user using the first *CONNECT* frame they send. In that frame, the user will send headers with their credentials which we will then check. Sending the credentials in the header is not the most secure way of doing this but it can give you ideas for your own system. For example, you can find a way to send a [JWT token](https://jwt.io) in the header. The way we are going to intercept the frames is with an implementation of the `ChannelListener` interface. As the name suggests, it defines a class you can use to intercept frames.
+
+### Implementing WebSocket Security
+
+First, we need to add the *spring-boot-starter-security* dependency to our project. Then, we need to set up a `User` entity and implement `UserDetailsService`. To focus on the topic of this guide we will skip over that. Know that we will have a few classes in the background that I won't show you the source code for. These include a `User`, `UserService`, `AppUserDetailsService`, and an `AppUserDetails` class. Assuming you understand Spring Data and Spring security you should understand the purpose of these classes. Anyways, next, we need to configure Spring security for our application:
```

Comment:
Please give a brief explanation for the User, UserService, AppUserDetailsService and AppUserDetails classes (a one line explanation suffices)

Reply:
Sure I can do that. I assumed they would be able to infer it if they understood the typical design patterns but I guess it's better safe than sorry.

Reply:
Yep, just making sure that it is easy for a beginner to follow